src/color-extractor.ts

```diff
@@ -0,0 +1,418 @@
+        break;
+      }
+      oldCenters = centers.slice();
+    }
+    return [centers, centerCounts, centerToColor];
+  }
+
+  /**
+   * Randomly initializes centers for k-means algorithm. Chooses k random 
+   * colors from an array of possible colors to be intial centers.
+   */
+  private initializeCenters(numClusters: number): Array<string> {
+    const centerCandidates = this.getCenterCandidates();
+    if (numClusters >= centerCandidates.length) {
+      return centerCandidates;
+    }
+    let centers = new Array<string>();
+    let usedIndices = new Set<number>();
+    while (centers.length < numClusters) {
+      let index = Math.floor(Math.random() * centerCandidates.length);
+      while (usedIndices.has(index)) {
+        index = Math.floor(Math.random() * centerCandidates.length);
+      }
+      centers.push(centerCandidates[index]);
+      usedIndices.add(index);
+    }
+    return centers;
+  }
+
+  /**
+   * Gets center candidates for performing k means on the data.
+   * Eliminates candidates if the count for a certain color is too low.
+   */
+  private getCenterCandidates(): Array<string> {
+    let centerCandidates = new Array<string>();
+    const threshold = this.pixelCount / this.keys.length / 2 ;
+    for (let i = 0; i < this.keys.length; i++) {
+      const count = this.data.get(this.keys[i]);
+      if (count && count > threshold){
+        centerCandidates.push(this.keys[i]);
+      }
+    }
+    return centerCandidates;
+  }
+
+  /**
+   * Updates centers for k means. First reassigns pixels to new clusters,
+   * then recalculates cluster centers based on reassignments. Returns
+   * new centers, counts for new centers, and a map from each center to
+   * an array of the colors assigned to that center's cluster.
+   * 
+   * @param centers   Array of old centers.
+   */
+  private updateCenters(centers: Array<string>): Array<any> {
+
+    // calculate new clusters
+    const pointToCenter = this.getNewClusters(centers);
+
+    // calculate new centers
+    const [centerCounts, centerSums] = 
+      this.getCenterStats(centers.length, pointToCenter);
+    const newCenters = this.getNewCenters(centerCounts, centerSums);
+
+    // generate map from center to all colors in cluster
+    let centerToColor: Map<string, Array<string>> = 
+      new Map<string, Array<string>>();
+    for (let i = 0; i < newCenters.length; i++) {
+      centerToColor.set(newCenters[i], []);
+    }
+    for (let i = 0; i < this.keys.length; i++) {
+      centerToColor.get(newCenters[pointToCenter[i]])?.push(this.keys[i]);
+    }
+
+    return [newCenters, centerCounts, centerToColor];
+  }
+
+  /**
+   * Gets new cluster assignments based on colors and centers.
+   * 
+   * @param centers  Array of current centers.
+   */
+  private getNewClusters(centers: Array<string>): Array<number> {
+    let pointToCenter: Array<number> = [];
+    for (let i = 0; i < this.keys.length; i++) {
+      let minDist = 100;
```

Comment:
Is 100 an arbitrary number? Is it big enough?

Reply:
100 is pretty much the largest a lab color distance can be -- I tested this by finding the color distance between white and black.